File: frontend/src/types/target.ts
/**
 * Target scope enum - defines how far the crawler should explore from the target URL
 */
export const TargetScope = {
  /** Crawl only the specified domain */
  DOMAIN: "DOMAIN",
  /** Crawl domain and all subdomains */
  SUBDOMAIN: "SUBDOMAIN",
  /** Crawl only the exact URL provided */
  URL_ONLY: "URL_ONLY"
} as const;

/**
 * Target scope type derived from TargetScope values
 */
export type TargetScope = typeof TargetScope[keyof typeof TargetScope];

/**
 * Target entity returned from the backend API
 */
export interface Target {
  id: number;
  project_id: number;
  name: string;
  url: string;
  description: string | null;
  scope: TargetScope;
  created_at: string;
  updated_at: string;
}

/**
 * Data required to create a new target
 */
export interface TargetCreate {
  name: string;
  url: string;
  description?: string;
  scope?: TargetScope;
}

/**
 * Data required to update an existing target
 * All fields are optional
 */
export interface TargetUpdate {
  name?: string;
  url?: string;
  description?: string;
  scope?: TargetScope;
}

/**
 * Query parameters for fetching targets list
 */
export interface TargetListParams {
  skip?: number;
  limit?: number;
}

/**
 * Response returned when triggering a scan task
 */
export interface ScanTriggerResponse {
  status: string;
  task_id: number;
}

File: backend/app/models/target.py
from datetime import datetime
from typing import Optional
from sqlmodel import Field, SQLModel

from app.core.utils import utc_now

from enum import Enum


class TargetScope(str, Enum):
    DOMAIN = "DOMAIN"
    SUBDOMAIN = "SUBDOMAIN"
    URL_ONLY = "URL_ONLY"


class TargetBase(SQLModel):
    name: str = Field(index=True, max_length=255)
    url: str = Field(max_length=2048)
    description: Optional[str] = Field(default=None)
    scope: TargetScope = Field(default=TargetScope.DOMAIN)


class Target(TargetBase, table=True):
    __tablename__ = "targets"

    id: Optional[int] = Field(default=None, primary_key=True)
    project_id: int = Field(foreign_key="projects.id", nullable=False)
    created_at: datetime = Field(default_factory=utc_now)
    updated_at: datetime = Field(default_factory=utc_now)


class TargetCreate(TargetBase):
    pass


class TargetUpdate(SQLModel):
    name: Optional[str] = None
    url: Optional[str] = None
    description: Optional[str] = None
    scope: Optional[TargetScope] = None


class TargetRead(TargetBase):
    id: int
    project_id: int
    created_at: datetime
    updated_at: datetime
    asset_count: int = 0
